src/middlewared/middlewared/plugins/datastore/connection.py

```diff
@@ -27,6 +29,19 @@ class Config:
     engine = None
     connection = None
 
+    @private
+    def handle_constraint_violation(self, row):
+        self.logger.warning("Row %d in table %s violates foreign key constraint on table %s.",
+                            row.rowid, row.table, row.parent)
+
+        if row.table == "directoryservice_idmap_domain" and row.rowid <= DSType.DS_TYPE_DEFAULT_DOMAIN.value:
```

Comment:
We should probably check `row.parent` also to ensure that the violating column is indeed `idmap_domain_certificate_id` and not some other column

Reply:
Isnt this code violating the database layer and the database should be extended so the logic is on idmap side?